### PR TITLE
Replace InstancesDao with InstancesRepository

### PR DESCRIPTION
**Pull request by seadowg**
Work towards #4454. There are still a few calls to that touch Instances going straight through the Content Resolver but this will be easier to clean up once we've go rid of calls to `FormsDao` as well.

I'd hope this pretty reviewable commit by commit as I tried to keep each step pretty darn small for my own sanity.

#### What has been done to verify that this works as intended?

New tests and verified manually.

#### Why is this the best possible solution? Were any other approaches considered?

This PR is pretty rinse and repeat in its approach:

1. Find a Dao call
2. Work out if the code that it's part of is tested and try and add one if not
3. Add a new method to `InstancesRepository` if needed and then replace the dao call with one to the repo

I'll leave comments inline on any other changes I had to make.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

I've tried to add tests for the core parts of everything I've changed but I of course could have missed things. I think it's worth waiting to swap out the forms calls as well before doing anything with QA but it feels like it might be worth doing something close to a regression pass (focused on forms/instances management) at that time.

For reviewers, I think it'd be good to look out for anywhere where the replacement is actually doing something different from the original dao call. Additionally, it would be good to think about if any changes have negative performance impacts.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/QRCodePage.java [hunk 1]
```diff
@@ -39,7 +40,13 @@ public QRCodePage clickScanFragment() {
     }
 
     public QRCodePage clickView() {
-        onView(withText(R.string.view_qr_code_fragment_title)).perform(click());
+        // Switching tabs doesn't seem to work sometimes
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/pages/QRCodePage.java:43]: This kept causing me problems in my test lab runs.

File: collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
+package org.odk.collect.android.dao;
+
+import androidx.loader.content.CursorLoader;
+
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.instances.Instance;
+import org.odk.collect.android.provider.InstanceProviderAPI;
+
+public class CursorLoaderFactory {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java:9]: The plan for the moment is to move all the `CursorLoader` calls here. Removing `CursorLoader` is probably going to be a job for sometime in the future so having them located in one easy to find place will hopefully make this easier.

File: collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java [hunk 3]
```diff
@@ -60,68 +53,25 @@ protected void onPostExecute(Outcome outcome) {
...
+
+                    // The custom configuration from the third party app overrides
+                    // the app preferences set for delete after submission
+                    if (deleteInstanceAfterSubmission != null) {
+                        isFormAutoDeleteOptionEnabled = deleteInstanceAfterSubmission;
+                    } else {
+                        isFormAutoDeleteOptionEnabled = settingsProvider.getGeneralSettings().getBoolean(GeneralKeys.KEY_DELETE_AFTER_SEND);
                     }
+
+                    Stream<Instance> instancesToDelete = instanceIds.stream()
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java:68]: This is less efficient in terms of the amount of I/O performed as it makes a call for each instance that has just been submitted. I did think about optimizing this but realized:

* in most cases the instance count will be fairly low and so not a big deal
* if the instance count is large we'd actually prefer to do a query per instance rather than load them all into memory

Given that I think this is fine. If we really want we could add a way to grab several Instances by id but I suspect we could optimize this by holding onto instance refs we supposedly load during submission.

File: collect_app/src/main/java/org/odk/collect/android/upload/InstanceUploader.java [hunk 4]
```diff
@@ -49,55 +42,26 @@
...
-            selectionBuf.append(')');
-            String selection = selectionBuf.toString();
+        List<Instance> instances = new ArrayList<>();
 
-            Cursor c = dao.getInstancesCursor(selection, selectionArgs);
-            instancesToUpload.addAll(dao.getInstancesFromCursor(c));
-
-            counter++;
+        for (Long id : instanceDatabaseIds) {
+            instances.add(new DatabaseInstancesRepository().get(id));
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/upload/InstanceUploader.java:48]: Again this could be more efficient but not sure it's a big deal given the instance count will usually be low.

File: collect_app/src/test/java/org/odk/collect/android/tasks/InstanceSyncTaskTest.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.File;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+@RunWith(AndroidJUnit4.class)
+public class InstanceSyncTaskTest {
+
+    @Test
+    public void whenAnInstanceFileNoLongerExists_deletesFromDatabase() {
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/tasks/InstanceSyncTaskTest.java:22]: I mostly wrote this test to prove to myself what was happening.

File: collect_app/src/main/java/org/odk/collect/android/dao/helpers/InstancesDaoHelper.java [hunk 6]
```diff
@@ -74,15 +74,12 @@ public static boolean isInstanceComplete(boolean end, boolean completedByDefault
...
         if (path != null) {
-            try (Cursor c = new InstancesDao().getInstancesCursorForFilePath(path)) {
-                if (c != null && c.getCount() > 0) {
-                    // should only be one...
-                    c.moveToFirst();
-                    String id = c.getString(c.getColumnIndex(InstanceColumns._ID));
-                    return Uri.withAppendedPath(InstanceColumns.CONTENT_URI, id);
-                }
+            Instance instance = new DatabaseInstancesRepository().getOneByPath(path);
+            if (instance != null && instance.getStatus().equals(Instance.STATUS_COMPLETE)) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/dao/helpers/InstancesDaoHelper.java]: Why did you add this additional condition here (`&& instance.getStatus().equals(Instance.STATUS_COMPLETE)`)

File: collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java (hunk 2 above)
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java:9]: Don't we need a test for this class?

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java [hunk 7]
```diff
@@ -93,7 +93,7 @@ private void init() {
         setupAdapter();
         instanceSyncTask = new InstanceSyncTask(settingsProvider);
         instanceSyncTask.setDiskSyncListener(this);
-        instanceSyncTask.execute();
+//        instanceSyncTask.execute();
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java:96]: What's the next step here after disabling the task?

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]
Looks good!

File: collect_app/src/main/java/org/odk/collect/android/dao/helpers/InstancesDaoHelper.java (hunk 6 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/dao/helpers/InstancesDaoHelper.java]: Hmmm looks like a mistake. I feel like that should probably break something, but it doesn't seem to. I'll look into that.

File: collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java (hunk 2 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java:9]: The code is being called from the same places but have now just moved into a new object. As far as I can recall there was no testing for the individual `CursorLoader` methods before, but they're all used as UI so should have a lot of implicit coverage from our Espresso tests 🤞. Is there any logic in particular here you felt might need a test?

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java (hunk 7 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java:96]: There's an issue filed [here](https://github.com/getodk/collect/issues/4474). Sorry I should have mentioned that in the PR description.

File: collect_app/src/main/java/org/odk/collect/android/dao/helpers/InstancesDaoHelper.java (hunk 6 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/dao/helpers/InstancesDaoHelper.java]: OK I've removed the extra code. Looked like it wasn't causing problems because it was only used in places where the instance would be completed.

File: collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java:9]: Since everything goes down to `getInstancesCursorLoader()` we could for example test that the method is called with proper parameters in all cases? but it's not a must I'm just thinking...

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java:9]: Hmmmmm yeah especially for things like ordering it'd be nice to know we're calling the right thing. I do think because none of this has changed (the methods have just moved) and it's all things we want to move I wouldn't look at backfilling tests as a priority. I think it'd be hard to write tests for the specifics (like sort order) that wouldn't have to be changed if we remove cursor loaders. Does that make sense?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/dao/CursorLoaderFactory.java:9]: Yeah

**Review by grzesiek2010 — APPROVED** [review]
